### PR TITLE
58 setup queries for nabla backend for campaign and mainnet reward program

### LINKED ISSUE #58: Setup queries for Nabla backend for campaign and mainnet reward program
## Context
As the Nabla testnet is soon would be ready, growth wants to do a airflyft campaign where they want to track the actions user performed during a certain period of time. After the testnet campaign and mainnet launch, we also want to introduce a reward program based on points, where we will track certain KPI and reward user with points. This is to attract LPs.

### Requirement
* Following queries should be available in the Nabla squid
* For testnet campaign, growth wants to track the following actions over a period of 1 week using airlyft
    * If the user did a swap
    * if the user added liquidity to the swap pools
    * If the user added liquidity to the backstop pool
* For the reward program, we want to track the following factors (as of now not decided but probable approach) 
    * Amount LPed (TVL)
    * Duration (for which the amount is LPed)
* Above queries should be available for both Foucoco and Pendulum

### Technical considerations

Currently, we have four entities for Nabla: `Router`, `NablaToken`, `SwapPool` and `BackstopPool`, see [here](https://github.com/pendulum-chain/pendulum-squids/blob/71e97fa96a0a4a272136a31a41312d861df41e5b/schema.graphql#L744-L777). We are not storing data about events like swapping or depositing liquidity. We should:

- Create a new entity `NablaSwap` based on the [Swap](https://github.com/pendulum-chain/pendulum-squids/blob/71e97fa96a0a4a272136a31a41312d861df41e5b/schema.graphql#L475) entity we use already for Zenlink. 

- Create new entities `NablaBackstopLiquidityDeposit` and `NablaSwapLiquidityDeposit` based on the [Mint](https://github.com/pendulum-chain/pendulum-squids/blob/71e97fa96a0a4a272136a31a41312d861df41e5b/schema.graphql#L424-L446) entity
- Create new entities `NablaBackstopLiquidityWithdrawal` and `NablaSwapLiquidityWithdrawal` based on the [Burn](https://github.com/pendulum-chain/pendulum-squids/blob/71e97fa96a0a4a272136a31a41312d861df41e5b/schema.graphql#L448-L473) entity

The new entities should store the relevant general metadata like `id`, `transaction`, `timestamp` and `pair` and then take the other data from the respective event emitted and processed by the indexer. The available events can be seen in the `abi` folder [here](https://github.com/pendulum-chain/pendulum-squids/blob/main/src/abi/swap.ts) and [here](https://github.com/pendulum-chain/pendulum-squids/blob/main/src/abi/backstop.ts).

To derive the values for 'Amount LPed (TVL)' and 'Duration (for which the amount is LPed)' we would then manually (or with a script) compare the `NablaXDeposit` and `NablaXWithdrawal` events for each user to be able to say how many tokens were provided as liquidity and for how long.

### PR DESCRIPTION
**Pull request by bogdanS98**
Closes #58.

This should be merged to `main` only after #52 is merged.

Tested some of these changes locally by running the nabla indexer tests from [this](https://github.com/pendulum-chain/wasm-deploy/pull/40) `wasm-deploy` branch, therefore generating some events data. This is not enough, so more events must be generated in order to test all queries.

### CODE REVIEW

**Comment by prayagd** [comment]
@pendulum-chain/devs can we please review this? it would unblock the nabla testnet campaign

File: schema.graphql
Comment (TorstenStueber) [inline schema.graphql]: The notion `principle` is taken from the Nabla event definition. "Principle" was the old word the Nabla team used for pool token. This is still used in the `Mint` and `Burn` event definitions.

You might want to call this `amountPoolTokensDeposited/...Withdrawn` in this and the following entity definitions instead.

**Review by TorstenStueber — DISMISSED** [review]
Just one minor comment. I already approve and leave it to you to change the name if this is low effort.

File: src/mappings/nabla/creation.ts
Comment (ebma) [inline src/mappings/nabla/creation.ts]: We are repeatedly using the same logic to derive the id of these items. Let's rather extract that into a function and call that function. This way, it's less error prone and easy to change in the future.

Comment (ebma) [reply src/mappings/nabla/creation.ts]: I'm not a big fan of the ID derivation in general. How about we always make it a combination of the related item (in this example, the backstop pool) + e.g. `ctx.event.id`or similar. Then we also have to pass two arguments less into each of these calls.

File: schema.graphql
Comment (ebma) [inline schema.graphql]: Maybe we can hook these up to our existing [NablaToken](https://github.com/pendulum-chain/pendulum-squids/blob/fe71b51ae6a6dfa35e1c66a3b22ef73400d485cc/schema.graphql#L751) entity?

**Review by ebma — CHANGES_REQUESTED** [review]
I would expect the sender to have a properly encoded address so that I can easily look up the account on chain. This is not the case at the moment.
image

I also get quite a lot of errors when processing on Foucoco locally. Any ideas @bogdanS98?
```
18:08:08 INFO  sqd:processor 1508390 / 1562149, rate: 1449 blocks/sec, mapping: 939 blocks/sec, 1681 items/sec, eta: 37s
Error processing event 'Contracts.ContractEmitted'. Skipping due to error: QueryFailedError: update or delete on table "nabla_swap_fee" violates foreign key constraint "FK_dd24afed240ad46290bfcf0603b" on table "nabla_swap"
    at PostgresQueryRunner.query (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:211:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async DeleteQueryBuilder.execute (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/DeleteQueryBuilder.js:52:33)
    at async Store.remove (/Users/marcel/Documents/pendulum-squids/node_modules/@subsquid/typeorm-store/lib/store.js:114:17)
    at async filterSwapFeeHistory (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/swapPoolEventHandler.js:118:13)
    at async updateAndPruneSwapFeeHistory (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/swapPoolEventHandler.js:112:5)
    at async updateAprAfterSwap (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/swapPoolEventHandler.js:124:5)
    at async handleChargedSwapFees (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/swapPoolEventHandler.js:68:5)
    at async handleSwapPoolEvent (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/swapPoolEventHandler.js:23:13)
    at async handleContractEvent (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/handleEvent.js:71:13) {
  query: 'DELETE FROM "nabla_swap_fee" WHERE "id" IN ($1)',
  parameters: [ '1495223-3' ],
  driverError: error: update or delete on table "nabla_swap_fee" violates foreign key constraint "FK_dd24afed240ad46290bfcf0603b" on table "nabla_swap"
      at Parser.parseErrorMessage (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:287:98)
      at Parser.handlePacket (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:126:29)
      at Parser.parse (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:39:38)
      at Socket.<anonymous> (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/index.js:11:42)
      at Socket.emit (node:events:512:28)
      at addChunk (node:internal/streams/readable:324:12)
      at readableAddChunk (node:internal/streams/readable:297:9)
      at Readable.push (node:internal/streams/readable:234:10)
      at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
    length: 317,
    severity: 'ERROR',
    code: '23503',
    detail: 'Key (id)=(1495223-3) is still referenced from table "nabla_swap".',
    hint: undefined,
    position: undefined,
    internalPosition: undefined,
    internalQuery: undefined,
    where: undefined,
    schema: 'public',
    table: 'nabla_swap',
    column: undefined,
    dataType: undefined,
    constraint: 'FK_dd24afed240ad46290bfcf0603b',
    file: 'ri_triggers.c',
    line: '2609',
    routine: 'ri_ReportViolation'
  },
  length: 317,
  severity: 'ERROR',
  code: '23503',
  detail: 'Key (id)=(1495223-3) is still referenced from table "nabla_swap".',
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: 'public',
  table: 'nabla_swap',
  column: undefined,
  dataType: undefined,
  constraint: 'FK_dd24afed240ad46290bfcf0603b',
  file: 'ri_triggers.c',
  line: '2609',
  routine: 'ri_ReportViolation'
}
Error processing event 'Tokens.Transfer'. Skipping due to error: QueryFailedError: current transaction is aborted, commands ignored until end of transaction block
    at PostgresQueryRunner.query (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:211:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async InsertQueryBuilder.execute (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/InsertQueryBuilder.js:106:33)
    at async Store.upsert (/Users/marcel/Documents/pendulum-squids/node_modules/@subsquid/typeorm-store/lib/store.js:38:13)
    at async handleTokenTransfer (/Users/marcel/Documents/pendulum-squids/lib/mappings/token.js:329:5)
    at async /Users/marcel/Documents/pendulum-squids/lib/processor.js:130:25
    at async TypeormDatabase.performUpdates (/Users/marcel/Documents/pendulum-squids/node_modules/@subsquid/typeorm-store/lib/database.js:139:13)
    at async /Users/marcel/Documents/pendulum-squids/node_modules/@subsquid/typeorm-store/lib/database.js:84:13
    at async EntityManager.transaction (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/entity-manager/EntityManager.js:73:28)
    at async TypeormDatabase.submit (/Users/marcel/Documents/pendulum-squids/node_modules/@subsquid/typeorm-store/lib/database.js:151:24) {
  query: 'INSERT INTO "token_transfer"("id", "block_number", "timestamp", "extrinsic_hash", "from", "to", "currency_id", "amount", "remark") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, DEFAULT) ON CONFLICT ( "id" ) DO UPDATE SET "id" = EXCLUDED."id", "block_number" = EXCLUDED."block_number", "timestamp" = EXCLUDED."timestamp", "extrinsic_hash" = EXCLUDED."extrinsic_hash", "from" = EXCLUDED."from", "to" = EXCLUDED."to", "currency_id" = EXCLUDED."currency_id", "amount" = EXCLUDED."amount"',
  parameters: [
    '0001522352-781cd-000031',
    1522352,
    2024-05-08T11:15:48.155Z,
    '0x279944f9ff33020065859b585b89b63c2bff00c6641e1e67d314d4266e3c3a8f',
    '6kjpZPe7tkAKD7e8UyLeLMz1vUEDgkwxXuyCrtiCWXUzhe7u',
    '6ijJtaZuwpZCiaVo6pSHRJbd8qejgywYsejnjfo2AVanN14E',
    'XCM(0)',
    '904545686'
  ],
  driverError: error: current transaction is aborted, commands ignored until end of transaction block
      at Parser.parseErrorMessage (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:287:98)
      at Parser.handlePacket (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:126:29)
      at Parser.parse (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:39:38)
      at Socket.<anonymous> (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/index.js:11:42)
      at Socket.emit (node:events:512:28)
      at addChunk (node:internal/streams/readable:324:12)
      at readableAddChunk (node:internal/streams/readable:297:9)
      at Readable.push (node:internal/streams/readable:234:10)
      at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
    length: 145,
    severity: 'ERROR',
    code: '25P02',
    detail: undefined,
    hint: undefined,
    position: undefined,
    internalPosition: undefined,
    internalQuery: undefined,
    where: undefined,
    schema: undefined,
    table: undefined,
    column: undefined,
    dataType: undefined,
    constraint: undefined,
    file: 'postgres.c',
    line: '1470',
    routine: 'exec_parse_message'
  },
  length: 145,
  severity: 'ERROR',
  code: '25P02',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'postgres.c',
  line: '1470',
  routine: 'exec_parse_message'
}
Error processing event 'Contracts.ContractEmitted'. Skipping due to error: QueryFailedError: current transaction is aborted, commands ignored until end of transaction block
    at PostgresQueryRunner.query (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:211:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async SelectQueryBuilder.loadRawResults (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/SelectQueryBuilder.js:2183:25)
    at async SelectQueryBuilder.getRawMany (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/SelectQueryBuilder.js:646:29)
    at async SelectQueryBuilder.executeEntitiesAndRawResults (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/SelectQueryBuilder.js:1973:26)
    at async SelectQueryBuilder.getRawAndEntities (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/SelectQueryBuilder.js:684:29)
    at async SelectQueryBuilder.getOne (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/SelectQueryBuilder.js:711:25)
    at async getBackstopPool (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/creation.js:34:12)
    at async Promise.all (index 0)
    at async handleContractEvent (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/handleEvent.js:58:61) {
```

**Comment by TorstenStueber** [comment]
Good catch @ebma! @bogdanS98 you need to be careful to cleanly distinguish addresses encoded in hex and ss58 format. Both are of type `string`, so TypeScript doesn't help. You will see that I tried to always add the postfix `hexAddress` or `ss58Address` to variable names (e.g. [here](https://github.com/pendulum-chain/pendulum-squids/blob/cb9ab7ecb077d4df6a74499d7eb0ad541918730b/src/mappings/nabla/creation.ts#L61)) to at least track it using the name.

Note:
* all entries in the database (particularly ids) should use the ss58 encoding
* the events the indexer handles will always contain the hex address

File: src/mappings/nabla/creation.ts
Comment (TorstenStueber) [reply src/mappings/nabla/creation.ts]: A possible argument could be, e.g., [this one](https://github.com/pendulum-chain/pendulum-squids/pull/62/files#diff-0f12e301645d6153e996e498d2c74df4bc64af81718df3ab0f9d46a4e814b0daR95): here we need to request some entry from the `NablaSwapFee` table that has been created in another event shortly before – the function `swapExactTokensForTokens` of the router contract will generate the following actions:
1. the swap out pool emits the `ChargedSwapFees` event, which will be handled [in the indexer](https://github.com/pendulum-chain/pendulum-squids/pull/62/files#diff-0f7e42b3ffe1b4451fc54f2546a359c89a0d16cae8347478995936b49e50fb6eR45) by creating a `NablaSwapFee` entry
1. the router emits a `Swap` event, the handler of which is the one I link to above

Using this id, it is easy to look up the associated `NablaSwapFee`.

A counter argument is that this would not work anymore for batch transactions.

**Comment by prayagd** [comment]
@pendulum-chain/devs any update on this? is there a fix to be done, can we please priortize this as this is blocking the campaign

**Comment by bogdanS98** [comment]
> @pendulum-chain/devs any update on this? is there a fix to be done, can we please priortize this as this is blocking the campaign

I'm working on this as we speak, will push the latest changes requested in the next few hours

**Comment by prayagd** [comment]
> I'm working on this as we speak, will push the latest changes requested in the next few hours

Thank you Bogdan for the update

**Comment by bogdanS98** [comment]
> I would expect the sender to have a properly encoded address so that I can easily look up the account on chain. This is not the case at the moment. image
> 
> I also get quite a lot of errors when processing on Foucoco locally. Any ideas @bogdanS98?
> 
> ```
> 18:08:08 INFO  sqd:processor 1508390 / 1562149, rate: 1449 blocks/sec, mapping: 939 blocks/sec, 1681 items/sec, eta: 37s
> Error processing event 'Contracts.ContractEmitted'. Skipping due to error: QueryFailedError: update or delete on table "nabla_swap_fee" violates foreign key constraint "FK_dd24afed240ad46290bfcf0603b" on table "nabla_swap"
>     at PostgresQueryRunner.query (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:211:19)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async DeleteQueryBuilder.execute (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/DeleteQueryBuilder.js:52:33)
>     at async Store.remove (/Users/marcel/Documents/pendulum-squids/node_modules/@subsquid/typeorm-store/lib/store.js:114:17)
>     at async filterSwapFeeHistory (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/swapPoolEventHandler.js:118:13)
>     at async updateAndPruneSwapFeeHistory (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/swapPoolEventHandler.js:112:5)
>     at async updateAprAfterSwap (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/swapPoolEventHandler.js:124:5)
>     at async handleChargedSwapFees (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/swapPoolEventHandler.js:68:5)
>     at async handleSwapPoolEvent (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/swapPoolEventHandler.js:23:13)
>     at async handleContractEvent (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/handleEvent.js:71:13) {
>   query: 'DELETE FROM "nabla_swap_fee" WHERE "id" IN ($1)',
>   parameters: [ '1495223-3' ],
>   driverError: error: update or delete on table "nabla_swap_fee" violates foreign key constraint "FK_dd24afed240ad46290bfcf0603b" on table "nabla_swap"
>       at Parser.parseErrorMessage (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:287:98)
>       at Parser.handlePacket (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:126:29)
>       at Parser.parse (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:39:38)
>       at Socket.<anonymous> (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/index.js:11:42)
>       at Socket.emit (node:events:512:28)
>       at addChunk (node:internal/streams/readable:324:12)
>       at readableAddChunk (node:internal/streams/readable:297:9)
>       at Readable.push (node:internal/streams/readable:234:10)
>       at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
>     length: 317,
>     severity: 'ERROR',
>     code: '23503',
>     detail: 'Key (id)=(1495223-3) is still referenced from table "nabla_swap".',
>     hint: undefined,
>     position: undefined,
>     internalPosition: undefined,
>     internalQuery: undefined,
>     where: undefined,
>     schema: 'public',
>     table: 'nabla_swap',
>     column: undefined,
>     dataType: undefined,
>     constraint: 'FK_dd24afed240ad46290bfcf0603b',
>     file: 'ri_triggers.c',
>     line: '2609',
>     routine: 'ri_ReportViolation'
>   },
>   length: 317,
>   severity: 'ERROR',
>   code: '23503',
>   detail: 'Key (id)=(1495223-3) is still referenced from table "nabla_swap".',
>   hint: undefined,
>   position: undefined,
>   internalPosition: undefined,
>   internalQuery: undefined,
>   where: undefined,
>   schema: 'public',
>   table: 'nabla_swap',
>   column: undefined,
>   dataType: undefined,
>   constraint: 'FK_dd24afed240ad46290bfcf0603b',
>   file: 'ri_triggers.c',
>   line: '2609',
>   routine: 'ri_ReportViolation'
> }
> Error processing event 'Tokens.Transfer'. Skipping due to error: QueryFailedError: current transaction is aborted, commands ignored until end of transaction block
>     at PostgresQueryRunner.query (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:211:19)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async InsertQueryBuilder.execute (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/InsertQueryBuilder.js:106:33)
>     at async Store.upsert (/Users/marcel/Documents/pendulum-squids/node_modules/@subsquid/typeorm-store/lib/store.js:38:13)
>     at async handleTokenTransfer (/Users/marcel/Documents/pendulum-squids/lib/mappings/token.js:329:5)
>     at async /Users/marcel/Documents/pendulum-squids/lib/processor.js:130:25
>     at async TypeormDatabase.performUpdates (/Users/marcel/Documents/pendulum-squids/node_modules/@subsquid/typeorm-store/lib/database.js:139:13)
>     at async /Users/marcel/Documents/pendulum-squids/node_modules/@subsquid/typeorm-store/lib/database.js:84:13
>     at async EntityManager.transaction (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/entity-manager/EntityManager.js:73:28)
>     at async TypeormDatabase.submit (/Users/marcel/Documents/pendulum-squids/node_modules/@subsquid/typeorm-store/lib/database.js:151:24) {
>   query: 'INSERT INTO "token_transfer"("id", "block_number", "timestamp", "extrinsic_hash", "from", "to", "currency_id", "amount", "remark") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, DEFAULT) ON CONFLICT ( "id" ) DO UPDATE SET "id" = EXCLUDED."id", "block_number" = EXCLUDED."block_number", "timestamp" = EXCLUDED."timestamp", "extrinsic_hash" = EXCLUDED."extrinsic_hash", "from" = EXCLUDED."from", "to" = EXCLUDED."to", "currency_id" = EXCLUDED."currency_id", "amount" = EXCLUDED."amount"',
>   parameters: [
>     '0001522352-781cd-000031',
>     1522352,
>     2024-05-08T11:15:48.155Z,
>     '0x279944f9ff33020065859b585b89b63c2bff00c6641e1e67d314d4266e3c3a8f',
>     '6kjpZPe7tkAKD7e8UyLeLMz1vUEDgkwxXuyCrtiCWXUzhe7u',
>     '6ijJtaZuwpZCiaVo6pSHRJbd8qejgywYsejnjfo2AVanN14E',
>     'XCM(0)',
>     '904545686'
>   ],
>   driverError: error: current transaction is aborted, commands ignored until end of transaction block
>       at Parser.parseErrorMessage (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:287:98)
>       at Parser.handlePacket (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:126:29)
>       at Parser.parse (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/parser.js:39:38)
>       at Socket.<anonymous> (/Users/marcel/Documents/pendulum-squids/node_modules/pg-protocol/dist/index.js:11:42)
>       at Socket.emit (node:events:512:28)
>       at addChunk (node:internal/streams/readable:324:12)
>       at readableAddChunk (node:internal/streams/readable:297:9)
>       at Readable.push (node:internal/streams/readable:234:10)
>       at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
>     length: 145,
>     severity: 'ERROR',
>     code: '25P02',
>     detail: undefined,
>     hint: undefined,
>     position: undefined,
>     internalPosition: undefined,
>     internalQuery: undefined,
>     where: undefined,
>     schema: undefined,
>     table: undefined,
>     column: undefined,
>     dataType: undefined,
>     constraint: undefined,
>     file: 'postgres.c',
>     line: '1470',
>     routine: 'exec_parse_message'
>   },
>   length: 145,
>   severity: 'ERROR',
>   code: '25P02',
>   detail: undefined,
>   hint: undefined,
>   position: undefined,
>   internalPosition: undefined,
>   internalQuery: undefined,
>   where: undefined,
>   schema: undefined,
>   table: undefined,
>   column: undefined,
>   dataType: undefined,
>   constraint: undefined,
>   file: 'postgres.c',
>   line: '1470',
>   routine: 'exec_parse_message'
> }
> Error processing event 'Contracts.ContractEmitted'. Skipping due to error: QueryFailedError: current transaction is aborted, commands ignored until end of transaction block
>     at PostgresQueryRunner.query (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:211:19)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async SelectQueryBuilder.loadRawResults (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/SelectQueryBuilder.js:2183:25)
>     at async SelectQueryBuilder.getRawMany (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/SelectQueryBuilder.js:646:29)
>     at async SelectQueryBuilder.executeEntitiesAndRawResults (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/SelectQueryBuilder.js:1973:26)
>     at async SelectQueryBuilder.getRawAndEntities (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/SelectQueryBuilder.js:684:29)
>     at async SelectQueryBuilder.getOne (/Users/marcel/Documents/pendulum-squids/node_modules/typeorm/query-builder/SelectQueryBuilder.js:711:25)
>     at async getBackstopPool (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/creation.js:34:12)
>     at async Promise.all (index 0)
>     at async handleContractEvent (/Users/marcel/Documents/pendulum-squids/lib/mappings/nabla/handleEvent.js:58:61) {
> ```

[This](https://github.com/pendulum-chain/pendulum-squids/pull/62/commits/75b2cc2af2f2689ec649fb1a2b15ba09bf09b0a4#diff-0f7e42b3ffe1b4451fc54f2546a359c89a0d16cae8347478995936b49e50fb6eR218-R227) fixes the errors in the console by setting the reference to `swapFee` to `null` before deleting a swap pool. 
Also, I converted hex addresses to ss58 👍🏼

File: src/mappings/nabla/creation.ts
Comment (bogdanS98) [reply src/mappings/nabla/creation.ts]: @ebma I extracted the id generation logic to a function, but I think we should create an extra ticket for [this](https://github.com/pendulum-chain/pendulum-squids/pull/62#discussion_r1611963997) since we have to take into account batch transactions as Torsten mentioned. WDYT? @ebma @TorstenStueber

File: schema.graphql
Comment (bogdanS98) [reply schema.graphql]: Done

File: schema.graphql
Comment (bogdanS98) [reply schema.graphql]: Done

**Review by ebma — COMMENTED** [review]
Can confirm that the error messages don't appear anymore during the processing, good job @bogdanS98 👍 

We should change the schema so that the deposit and withdrawal events link to either the respective swap or backstop pool. Otherwise, we are missing very important information and the events are useless.
image

File: src/mappings/nabla/creation.ts
Comment (TorstenStueber) [reply src/mappings/nabla/creation.ts]: It might be overkill to already think about batch transactions. In my view there is no ideal way to connect events that are generated during the same Nabla operation when they are actually batched in the same extrinsic. Or is there something that comes to your mind?

Maybe the best way is to avoid the situation all together and never try to connect events as we try to do here. Not sure yet what we can do instead.

File: src/mappings/nabla/backstopPoolEventHandler.ts
Comment (gianfra-t) [inline src/mappings/nabla/backstopPoolEventHandler.ts]: Shouldn't we call `createNablaBackstopLiquidityWithdrawal`? This one will save a `NablaSwapLiquidityWithdrawal` entity.

File: schema.graphql
Comment (ebma) [inline schema.graphql]: Nit: I think these comments are not needed as it's quite obvious that this backstop pool is the one related to the event. Same for the others.

**Review by ebma — APPROVED** [review]
LGTM

**Review by gianfra-t — APPROVED** [review]
LGTM too 👍

**Review by gianfra-t — APPROVED** [review]

**Review by ebma — APPROVED** [review]
Let's merge @bogdanS98?

**Pull request merged**
